### PR TITLE
Problem: No rotation/update of the solo machine keypair

### LINKED ISSUE #12: Problem: no rotation/update of the solo machine keypair
right now, it seems the keypair is fixed, but AFAIK one can broadcast a message to update the corresponding public key on the connection-channel?

### PR DESCRIPTION
**Pull request by devashishdxt**
Solution: Added CLI command and gRPC endpoint to update public key of solo machine client. Fixes #12.

### CODE REVIEW

File: solo-machine-core/src/service/ibc_service.rs [hunk 1]
```diff
@@ -417,6 +418,59 @@ impl IbcService {
             },
         )
     }
+
+    /// Updates signer for future IBC transactions
+    pub async fn update_signer(
+        &self,
+        signer: impl Signer,
```
Comment (tomtau) [inline solo-machine-core/src/service/ibc_service.rs:425]: should there also be a new signer, or can the signer trait handle both or one doesn't need to sign with the updated keypair?

**Review by tomtau — APPROVED** [review]
seems all right -- have you tested it on the latest ibc-go 1.0.0 that it works as expected?

File: solo-machine-core/src/event.rs [hunk 2]
```diff
@@ -2,11 +2,13 @@ use anyhow::{anyhow, Result};
 use tokio::sync::mpsc::UnboundedSender;
 
 use crate::{
+    cosmos::crypto::PublicKey,
     ibc::core::ics24_host::identifier::{ChainId, ChannelId, ClientId, ConnectionId, Identifier},
     model::ConnectionDetails,
 };
 
 /// Events emitted by IBC service
+#[allow(clippy::large_enum_variant)]
```
[the remark this answers is not in the document]
Comment (devashishdxt) [reply solo-machine-core/src/event.rs:11]: I don't see any huge performance impact. We can box it later if it becomes more complex.

File: solo-machine-core/migrations/20210717070615_chain_keys.up.sql [hunk 3]
```diff
@@ -0,0 +1,8 @@
+CREATE TABLE IF NOT EXISTS chain_keys (
+    id INTEGER PRIMARY KEY AUTOINCREMENT,
+    chain_id TEXT NOT NULL,
+    public_key TEXT NOT NULL,
+    created_at DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP,
+    UNIQUE(chain_id, public_key)
```
[the remark this answers is not in the document]
Comment (devashishdxt) [reply solo-machine-core/migrations/20210717070615_chain_keys.up.sql:6]: Right now, API does not allow changing diversifier. I don't clearly understand the significance of diversifier. Do you think it is important and user should be able to change that?

File: solo-machine-core/src/service/ibc_service.rs (hunk 1 above)
Comment (devashishdxt) [reply solo-machine-core/src/service/ibc_service.rs:425]: The transaction to update the public key is signer by old signer. All the future transactions will be signed by the new signer.

**Comment by devashishdxt** [comment]
> have you tested it on the latest ibc-go 1.0.0 that it works as expected?

I've tested with `ibc-go v1.0.0-beta1`. Will test it with `rc1` today.

File: solo-machine-core/migrations/20210717070615_chain_keys.up.sql (hunk 3 above)
[the remark this answers is not in the document]
Comment (tomtau) [reply solo-machine-core/migrations/20210717070615_chain_keys.up.sql:6]: my understanding of the diversifier is that it's if one wanted to use the same solo machine keypair on multiple channels/chains... so I guess it's ok if it's set once in the beginning for each chain, i.e. ok to keep the "set once in the initial config" for each chain now?
perhaps the only use case may be if one made a mistake and wanted to update it

**Pull request merged**